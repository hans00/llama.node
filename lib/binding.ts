import * as path from 'path'

export type LlamaModelOptions = {
  model: string
  embedding?: boolean
  n_ctx?: number
  n_batch?: number
  n_threads?: number
  n_gpu_layers?: number
  use_mlock?: boolean
  use_mmap?: boolean
}

export type LlamaCompletionOptions = {
  prompt: string
  n_samples?: number
  temperature?: number
  top_k?: number
  top_p?: number
  repetition_penalty?: number
  n_predict?: number
  max_length?: number
  max_tokens?: number
  seed?: number
  stop?: string[]
}

export type LlamaCompletionResult = {
  text: string
  tokens_predicted: number
  tokens_evaluated: number
  truncated: boolean
}

export type LlamaCompletionToken = {
  token: string
}

export interface LlamaContext {
  new (options: LlamaModelOptions): LlamaContext
  getSystemInfo(): string
  completion(options: LlamaCompletionOptions, callback?: (token: LlamaCompletionToken) => void): Promise<LlamaCompletionResult>
  stopCompletion(): void
  saveSession(path: string): Promise<void>
  loadSession(path: string): Promise<void>
  release(): Promise<void>
}

export interface Module {
  LlamaContext: LlamaContext
}

export type LibVariant = 'default' | 'opencl'

const setupEnv = (variant?: string) => {
  const postfix = variant ? `-${variant}` : ''
  const binPath = path.resolve(__dirname, `../bin/${process.platform}${postfix}/${process.arch}/`)
  const systemPathEnv = process.env.PATH ?? process.env.Path ?? ''
  if (!systemPathEnv.includes(binPath)) {
    if (process.platform === 'win32') {
      process.env.Path = `${binPath};${systemPathEnv}`
    } else {
      process.env.PATH = `${binPath}:${systemPathEnv}`
    }
  }
}

export const loadModule = async (variant?: LibVariant): Promise<Module> => {
  try {
    if (variant && variant !== 'default') {
      setupEnv(variant)
      return await import(`../bin/${process.platform}-${variant}/${process.arch}/llama-node.node`) as Module
    }
  } catch {} // ignore errors and try the common path
  setupEnv()
  return await import(`../bin/${process.platform}/${process.arch}/llama-node.node`) as Module
}