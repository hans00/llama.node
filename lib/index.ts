import { loadModule, LlamaModelOptions } from './binding'
import type { Module, LlamaContext } from './binding'

export * from './binding'

export interface LlamaModelOptionsExtended extends LlamaModelOptions {
  lib_variant?: string
}

let module: Module | null = null

export const loadModel = async (options: LlamaModelOptionsExtended): Promise<LlamaContext> => {
    module ??= await loadModule(options.lib_variant)
    return new module.LlamaContext(options)
}
