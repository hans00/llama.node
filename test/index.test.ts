import path from 'path'
import waitForExpect from 'wait-for-expect'
import { loadModel } from '../lib'

it('work fine', async () => {
  let tokens = ''
  const model = await loadModel({ model: path.resolve(__dirname, './tiny-random-llama.gguf') })
  const result = await model.completion({
    prompt: 'My name is Merve and my favorite',
    n_samples: 1,
    temperature: 0,
    max_length: 100,
    n_predict: 10,
    seed: 0,
  }, (data) => {
    expect(data).toMatchObject({ token: expect.any(String) })
    tokens += data.token
  })
  expect(result).toMatchSnapshot()
  await waitForExpect(() => {
    expect(tokens).toBe(result.text)
  })
  await model.release()
})
