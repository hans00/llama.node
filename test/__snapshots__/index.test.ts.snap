// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`work fine 1`] = `
{
  "text": " swochadoorter scientific WindowsCa occupiedr√• alta",
  "tokens_evaluated": 18,
  "tokens_predicted": 10,
  "truncated": false,
}
`;
